### PR TITLE
Add conformace tests for `list-constructor-expr`

### PR DESCRIPTION
**Pull request by lochana-chathura**
## Purpose
$subject.

Fixes ballerina-platform/ballerina-lang#32637

### CODE REVIEW

File: conformance/lang/expressions/list_constructor.balt [hunk 1]
```diff
@@ -0,0 +1,1242 @@
...
+Labels: module-const-decl, list-constructor-expr, is-expr, tuple-type, readonly-type, string, boolean
+
+const [int, string, boolean] tuple1 = [11, "foo", true];
+const [int, [boolean, string], boolean] tuple2 = [11, [true, "foo"], true];
+
+function init() {
+    io:println(tuple1 is readonly); // @output true
+    io:println(tuple2 is readonly); // @output true
+}
+
```
Comment (pcnfernando) [inline conformance/lang/expressions/list_constructor.balt:42]: Shall we add a test to check during list construction each exp is being evaluated in order?
> The members of the list come from evaluating each expression in the expr-list in order.

```ballerina
int[] arr = [foo(), foo()]; // foo returns incremented value per each invocation
```

File: conformance/lang/expressions/list_constructor.balt [hunk 2]
```diff
@@ -0,0 +1,1242 @@
...
+Labels: module-const-decl, list-constructor-expr, is-expr, array-type, readonly-type, string
+
+const string[] strArr = [];
+const int[] intArr = [1, 2, 3, 4];
+
+function init() {
+    io:println(strArr is readonly); // @output true
+    io:println(intArr is readonly); // @output true
+}
+
```
Comment (pcnfernando) [inline conformance/lang/expressions/list_constructor.balt:28]: Shall we add the simplest form in the block scope?

File: conformance/lang/expressions/list_constructor.balt [hunk 3]
```diff
@@ -0,0 +1,1242 @@
...
+    [any, any...] t2 = [11, 27, "foo", "bar"];
+    io:println(t2 is [any, any...]); // @output true
+    io:println(t2 is [any, any, any, any]); // @output false
+    io:println(t2 is [13, 25, "foo"]); // @output false
+}
+
+Test-Case: output
+Description: Test inherent type of the created list when the applicable contextually expected type
+             is a list-type-descriptor of anydata.
+Labels: list-constructor-expr, is-expr, anydata, union-type, array-type, string, string
```
Comment (pcnfernando) [inline conformance/lang/expressions/list_constructor.balt:108]:
```suggestion
Labels: list-constructor-expr, is-expr, anydata, union-type, array-type, string
```

File: conformance/lang/expressions/list_constructor.balt [hunk 4]
```diff
@@ -0,0 +1,1242 @@
...
+
+    io:println(t2 is [int...]); // @output true
+    io:println(t2 is [int, int, int]); // @output false
+    io:println(t2 is [byte...]); // @output false
+}
+
+Test-Case: output
+Description: Test 3 for list-constructor-expr in module-var-decl.
+Labels: module-init-var-decl, list-constructor-expr, is-expr, array-type, tuple-type, union-type, byte, boolean,
+        string, module-init-var-decl
```
Comment (pcnfernando) [inline conformance/lang/expressions/list_constructor.balt:791]:
```suggestion
Labels: module-init-var-decl, list-constructor-expr, is-expr, array-type, tuple-type, union-type, byte, boolean,
        string
```

File: conformance/lang/expressions/list_constructor.balt [hunk 5]
```diff
@@ -0,0 +1,1311 @@
...
+
+    any|any[] a5 = [1, 2, 3];
+    io:println(a5 is (any|error)[]); // @output true
+    io:println(a5 is any[]); // @output false
+    io:println(a5 is (any|error)[3]); // @output false
+}
+
+Test-Case: output
+Description: Test inherent type of the created list when the applicable contextually expected type
+             is a list-type-descriptor with any members.
```
Comment (MaryamZi) [inline conformance/lang/expressions/list_constructor.balt:156]: We can be more specific here an say tuple type-descriptor, right?

File: conformance/lang/expressions/list_constructor.balt [hunk 6]
```diff
@@ -0,0 +1,1311 @@
...
+
+    [any, any...] t2 = [11, 27, "foo", "bar"];
+    io:println(t2 is [any, any...]); // @output true
+    io:println(t2 is [any, any, any, any]); // @output false
+    io:println(t2 is [13, 25, "foo"]); // @output false
+}
+
+Test-Case: output
+Description: Test inherent type of the created list when the applicable contextually expected type
+             is a list-type-descriptor of anydata.
```
Comment (MaryamZi) [inline conformance/lang/expressions/list_constructor.balt:172]: And "array type-descriptor" here.

File: conformance/lang/expressions/list_constructor.balt [hunk 7]
```diff
@@ -0,0 +1,1311 @@
...
+
+    [json, json...] t2 = [true, 2, "foo"];
+    io:println(t2 is [json, json...]); // @output true
+    io:println(t2 is [json, json, json]); // @output false
+    io:println(t2 is [true, 2, "foo"]); // @output false
+}
+
+Test-Case: output
+Description: Test inherent type of the created list when the applicable contextually expected type
+             is a list-type-descriptor with different types.
```
Comment (MaryamZi) [inline conformance/lang/expressions/list_constructor.balt:254]: Didn't get what's "different" in the tests? We can't have different types with just arrays?

File: conformance/lang/expressions/list_constructor.balt [hunk 8]
```diff
@@ -0,0 +1,1311 @@
...
+    io:println(t1 is [int, string[2], boolean]); // @output true
+    io:println(t1 is [1, ["foo", "bar"], true]); // @output false
+
+    [int, [int, string...], int] t2 = [12, [2, "bar", "qux", "quux"], 21];
+    io:println(t2 is [int, [int, string...], int]); // @output true
+    io:println(t2 is [int, [int, string, string, string], int]); // @output false
+    io:println(t2 is [12, [2, "bar", "qux", "quux"], 21]); // @output false
+}
+
+Test-Case: output
```
Comment (MaryamZi) [inline conformance/lang/expressions/list_constructor.balt:364]: Shall we add some tests for when just `readonly` is the contextually-expected type also?

See https://github.com/ballerina-platform/ballerina-spec/issues/554.

File: conformance/lang/expressions/list_constructor.balt [hunk 9]
```diff
@@ -0,0 +1,1311 @@
...
+Description: Test when there is no contextually expected type, then the inherent type is a a tuple-type-descriptor
+             with a member-type-descriptor for each expression in the expr-list,
+             where the type of each member-type-descriptor will be the broad type of the corresponding expression
+             in the expr-list.
+Labels: list-constructor-expr, is-expr, var, boolean-literal, singleton-type, DecimalNumber, tuple-type, string
+
+function init() {
+    var a1 = [1, 2, 3];
+    io:println(a1 is [int, int, int]); // @output true
+    io:println(a1 is [byte, byte, byte]); // @output false
```
Comment (MaryamZi) [inline conformance/lang/expressions/list_constructor.balt:718]: Shall we also test
```suggestion
    io:println(a1 is [1, 2, 3]); // @output false
```

File: conformance/lang/expressions/list_constructor.balt [hunk 10]
```diff
@@ -0,0 +1,1311 @@
...
+}
+
+Test-Case: error
+Description: Test when there is a contextually expected type, then the type that the inherent type
+             requires for each list member provides the contextually expected type for the expression for the member.
+Labels: list-constructor-expr, member-access-expr, array-type, DecimalFloatingPointNumber, byte, float, union-type,
+        optional-type, unary-minus
+
+function errorFunction() {
+    byte[] a1 = [1, 2, 3];
```
Comment (MaryamZi) [inline conformance/lang/expressions/list_constructor.balt:745]: Contextually-expected type is more of a compile time concept. IMO, the tests for this sentence should be those having invalid member expressions in the list constructor.

```ballerina
function errorFunction() {
    int[] _ = [
        "", // @error expected 'int', found 'string'
        1.0 // @error expected 'int', found 'float'
    ];
}
```

File: conformance/lang/expressions/list_constructor.balt (hunk 7 above)
Comment (lochana-chathura) [reply conformance/lang/expressions/list_constructor.balt]: What I wanted to mean was something like "array-type-descriptor with different array member types". Updated the descriptions to make it obvious. Thanks!

**Comment by lochana-chathura** [comment]
Currently have 2 unknown labels:
1. Test raw-template-expr as list-constructor expressions ->  `object:RawTemplate` label
2. Test range-expr as list-constructor expressions. ->  `isolated-method` label

File: conformance/lang/expressions/list_constructor.balt [hunk 11]
```diff
@@ -0,0 +1,1374 @@
+Test-Case: parser-error
+Description: Test list constructor negative syntax.
```
Comment (MaryamZi) [inline conformance/lang/expressions/list_constructor.balt:2]: Maybe
```suggestion
Description: Test list constructor syntax errors.
```

File: conformance/lang/expressions/list_constructor.balt [hunk 12]
```diff
@@ -0,0 +1,1374 @@
...
+        3 // @error missing comma token
+    ];
+}
+
+Test-Case: output
+Description: Test basic list constructor expression without contextually expected type.
+Labels: list-constructor-expr, string
+
+function init() {
+    io:println([]); // @output []
```
Comment (MaryamZi) [inline conformance/lang/expressions/list_constructor.balt:20]: The description says "without a contextually expected type". But here, there is a contextually-expected type: the parameter type from `io:println()` - `io:Printable`.

No contextually-expected type is when used with `var`.

File: conformance/lang/expressions/list_constructor.balt [hunk 13]
```diff
@@ -0,0 +1,1379 @@
...
+
+Test-Case: output
+Description: Test when there is no contextually expected type, then the inherent type is a a tuple-type-descriptor
+             with a member-type-descriptor for each expression in the expr-list,
+             where the type of each member-type-descriptor will be the broad type of the corresponding expression
+             in the expr-list.
+Labels: boolean-literal, DecimalNumber, is-expr, list-constructor-expr, singleton-type, string, tuple-type, var
+
+function init() {
+    var a1 = [1, 2, 3];
```
Comment (MaryamZi) [inline conformance/lang/expressions/list_constructor.balt:777]: Are these the only inherent type tests we have with no contextually-expected type?

Can we add tests like

```ballerina
var a = [1, "foo", "bar", 3];
a[1] = 2; // error
```

to make sure it's `[int, string, string, int]` and not `(int|string)[4]`.

File: conformance/lang/expressions/list_constructor.balt [hunk 14]
```diff
@@ -0,0 +1,1379 @@
...
+        true, // @error incompatible types: expected 'int?', found 'boolean'
+        1.2, // @error incompatible types: expected 'int?', found 'float'
+        false // @error incompatible types: expected 'int?', found 'boolean'
+    ];
+
+    (byte|string)[] a3 = [
+        400, // @error incompatible types: expected '(byte|string)', found 'int'
+        1.2, // @error incompatible types: expected '(byte|string)', found 'float'
+        false // @error incompatible types: expected '(byte|string)', found 'boolean'
+    ];
```
Comment (pcnfernando) [inline conformance/lang/expressions/list_constructor.balt:826]: Shall we add one with a union case, with one list type?

File: conformance/lang/expressions/list_constructor.balt [hunk 15]
```diff
@@ -0,0 +1,1379 @@
...
+Labels: error-constructor-expr, error-type, is-expr, list-constructor-expr, trap-expr, tuple-type, union-type
+
+function init() {
+    io:println([trap 23, trap error("Oops!")] is [(int|error), error]); // @output true
+    io:println([trap 23, trap error("Oops!")] is [int, error]); // @output false
+}
+
+Test-Case: output
+Description: Test query-expr as list-constructor expressions.
+Labels: array-type, is-expr, list-constructor-expr, multiplicative-expr, query-expr, singleton-type, tuple-type
```
Comment (pcnfernando) [inline conformance/lang/expressions/list_constructor.balt:1341]: Test with list constructor expr at select as well?

File: conformance/lang/expressions/list_constructor.balt [hunk 16]
```diff
@@ -0,0 +1,1379 @@
...
+    ] is [int[], int[]]); // @output true
+
+    io:println([
+        from int i in [1, 2, 3]
+        select i * 2,
+        from int i in [1, 2, 3]
+        where i % 2 != 0
+        select i
+    ] is [[2, 4, 6], [1, 3]]); // @output false
+}
```
Comment (pcnfernando) [inline conformance/lang/expressions/list_constructor.balt:1359]: Test for list constructor with foreach?

File: conformance/lang/expressions/list_constructor.balt [hunk 17]
```diff
@@ -0,0 +1,1379 @@
...
+
+    io:println([1, ["foo", [true, 23]], [true, false, true]] is [int, [string, [boolean, int]], [boolean, boolean, boolean]]); // @output true
+    io:println([1, ["foo", [true, 23]], [true, false, true]] is [1, ["foo", [true, 23]], [true, false, true]]); // @output false
+}
+
+Test-Case: error
+Description: Test when there is a contextually expected type, then the type that the inherent type
+             requires for each list member provides the contextually expected type for the expression for the member,
+             for array.
+Labels: array-type, byte, DecimalFloatingPointNumber, float, list-constructor-expr, member-access-expr, optional-type,
```
Comment (rdulmina) [inline conformance/lang/expressions/list_constructor.balt:805]: `string`

File: conformance/lang/expressions/list_constructor.balt [hunk 18]
```diff
@@ -0,0 +1,1379 @@
...
+Description: Test raw-template-expr as list-constructor expressions.
+Labels: is-expr, list-constructor-expr, object:RawTemplate, raw-template-expr, tuple-type
+
+function init() {
+    io:println([`x + y = ${2 + 3}`, `Count:${1}`, `INSERT INTO Details VALUES (${"Foo"}, ${20})`] is [object:RawTemplate, object:RawTemplate, object:RawTemplate]); // @output true
+}
+
+Test-Case: output
+Description: Test structural-constructor-expr as list-constructor expressions.
+Labels: list-constructor-expr, mapping-constructor-expr, record-type, string, table-constructor-expr, table-type,
```
Comment (rdulmina) [inline conformance/lang/expressions/list_constructor.balt:982]: `key-specifier`, `readonly-field`

File: conformance/lang/expressions/list_constructor.balt [hunk 19]
```diff
@@ -0,0 +1,1379 @@
...
+                {id: 1, name: "Frank", age: 54},
+                {id: 2, name: "John", age: 47}
+            ],
+        {a: 1, b: "bar", c: true}
+    ] is [[int, string], table<record {|readonly int id; string name; int age;|}> key(id), record {|int a; string b; boolean c;|}]); // @output true
+}
+
+Test-Case: output
+Description: Test object-constructor-expr as list-constructor expressions.
+Labels: array-type, is-expr, list-constructor-expr, object-constructor-expr
```
Comment (rdulmina) [inline conformance/lang/expressions/list_constructor.balt:998]: `string`, `field-access-expr`

File: conformance/lang/expressions/list_constructor.balt [hunk 20]
```diff
@@ -0,0 +1,1459 @@
...
+function init() {
+    [int, byte, 3] t1 = [1, 2, 3];
+    io:println(t1 is [int, byte, 3]); // @output true
+    io:println(t1 is [int, 2, 3]); // @output false
+}
+
+Test-Case: output
+Description: Test inherent type of the created list when the applicable contextually expected type
+             is a array-type-descriptor of (any|error).
+Fail-Issue: ballerina-platform/ballerina-lang#33377
```
Comment (lasinicl) [inline conformance/lang/expressions/list_constructor.balt:143]: This issue has been fixed? 
But line number 166 has an `ambiguous type` error

Comment (lasinicl) [inline conformance/lang/expressions/list_constructor.balt:143]: This issue has been fixed? 
But line number 166 has an `ambiguous type` error

File: conformance/lang/expressions/list_constructor.balt [hunk 21]
```diff
@@ -0,0 +1,1459 @@
...
+Description: Test inherent type of the created list when the applicable contextually expected type
+             is a array-type-descriptor of user-defined type.
+Labels: array-type, DecimalNumber, is-expr, list-constructor-expr, module-type-defn, singleton-type, union-type
+
+type Ints 0|1|2|3;
+
+function init() {
+    Ints[] a1 = [1, 3, 0];
+    io:println(a1 is Ints[]); // @output true
+    io:println(a1 is (0|1|3)[]); // @output false
```
Comment (lasinicl) [inline conformance/lang/expressions/list_constructor.balt:555]: Shall we add `io:println(a1 is (0|1|2|3)[]); // @output true` also?

Comment (lasinicl) [inline conformance/lang/expressions/list_constructor.balt:555]: Shall we add `io:println(a1 is (0|1|2|3)[]); // @output true` also?

File: conformance/lang/expressions/list_constructor.balt (hunk 20 above)
Comment (lochana-chathura) [reply conformance/lang/expressions/list_constructor.balt]: enabled. can you check now?

**Review by pcnfernando — APPROVED** [review]

**Review by lasinicl — APPROVED** [review]

File: conformance/lang/expressions/list_constructor.balt [hunk 22]
```diff
@@ -0,0 +1,1379 @@
...
+
+function errorFunction() {
+    int[] arr1 = [; // @error missing close bracket token
+    int[] arr2 = ]; // @error missing open bracket token
+    int[] arr3 = [
+        1
+        2 // @error missing comma token
+        3 // @error missing comma token
+    ];
+}
```
[the remark this answers is not in the document]
Comment (lochana-chathura) [reply conformance/lang/expressions/list_constructor.balt:13]: Added with [c4cfa16](https://github.com/ballerina-platform/ballerina-spec/pull/1042/commits/c4cfa1618fb8fa8cbf9d2c3adc06e2186efcaa42)

**Review by pcnfernando — APPROVED** [review]

**Review by rdulmina — APPROVED** [review]

**Pull request merged**